README.md
# leicht
Telegram bot framework

![](https://api.travis-ci.org/Like-all/leicht.svg)
[ ![Download](https://api.bintray.com/packages/like-all/deb/leicht/images/download.svg) ](https://bintray.com/like-all/deb/leicht/_latestVersion)

### [man 1 leicht](https://github.com/Like-all/leicht/blob/master/src/man/leicht.1.md)


CHANGELOG.md
## CHANGELOG

---

draft (1.x)

  * Draft version roadmap
  * Moving from debian packages to docker images
  * Converting a manpage into a nice readme
  * Making it possible to configure the daemon using environment variables
  * Logging to stdout
  * TCP(HTTP?) incoming messages loop
  * Verifying incoming messages using permanent tokens
  * Issuing one time verification tokens

---

leicht (0.9-4)

  * Added socketMode config option
  * Testing also default values within config structure

 _Thu, 28 Jun 2018 01:03:45 -0700_

---

leicht (0.9-3)

  * (Hopefully) Finally fixed logging!

 _Thu, 28 Jun 2018 01:03:42 -0700_

---

leicht (0.9-2)

  * Fixed log messages routing

 _Fri, 01 Dec 2017 09:42:47 -0800_

---

leicht (0.9-1)

  * Refactored message logging

 _Fri, 01 Dec 2017 07:35:49 -0800_

---

leicht (0.8-4)

  * Revision bump

 _Fri, 01 Dec 2017 07:10:53 -0800_

---

leicht (0.8-3)

  * Stable build of 0.8

 _Sat, 29 Jul 2017 13:13:40 -0700_

---

leicht (0.8-2)

  * Build testing version also for jessie

 _Sat, 29 Jul 2017 12:56:26 -0700_

---

leicht (0.8-1)

  * Added doNotLogBlacklisted option

 _Fri, 28 Jul 2017 07:48:10 -0700_

---

leicht (0.7-5)

  * Fixed stable release

 _Fri, 28 Jul 2017 07:33:43 -0700_

---

leicht (0.7-4)

  * Stable release

 _Tue, 04 Jul 2017 02:30:44 -0700_

---

leicht (0.7-3)

  * First build for stretch

 _Sat, 24 Jun 2017 01:34:56 -0700_

---

leicht (0.7-2)

  * Fixed #8

 _Sat, 24 Jun 2017 01:23:46 -0700_

---

leicht (0.7-1)

  * Fixed issues/6

 _Mon, 24 Apr 2017 09:40:45 -0700_

---

leicht (0.6-7)

  * Update manpage

 _Mon, 24 Apr 2017 08:58:17 -0700_

---

leicht (0.6-6)

  * Upload build artifacts to Bintray

 _Mon, 09 Jan 2017 22:20:41 +0300_

---

leicht (0.6-5)

  * Pass direct URL of a file to the script

 _Fri, 06 Jan 2017 16:30:17 +0300_

---

leicht (0.6-4)

  * Append CfgParams.Script to logstring properly

 _Thu, 29 Dec 2016 14:29:35 +0300_

---

leicht (0.6-3)

  * Handle voice file id within script

 _Thu, 29 Dec 2016 13:00:15 +0300_

---

leicht (0.6-2)

  *  Check if Voice.FileID is not nil
  *  Check if update.Message.Voice is not nil

 _Thu, 29 Dec 2016 12:45:11 +0300_

---

leicht (0.6-1)

  * Handling voice files example
  * Fix the name of FileID field

 _Wed, 28 Dec 2016 23:05:31 +0300_

---

leicht (0.5-3)

  * v0.5 release 

 _Wed, 28 Dec 2016 22:45:03 +0300_

---

leicht (0.5-2)

  * Proper formatting of Chat ID

 _Wed, 12 Oct 2016 11:59:32 +0300_

---

leicht (0.5-1)

  * Write command string to the log

 _Wed, 12 Oct 2016 11:48:41 +0300_

---

leicht (0.4-2)

  * Fix integer base

 _Wed, 12 Oct 2016 11:30:38 +0300_

---

leicht (0.4-1)

  * Telegram bot library upgrade

 _Sat, 16 Jul 2016 16:49:46 +0300_

---

leicht (0.3-2)

  * v0.3 release

 _Sat, 16 Jul 2016 16:07:12 +0300_

---

leicht (0.3-1)

  * Improve log filenames

 _Wed, 18 May 2016 13:01:10 +0300_

---

leicht (0.2-7)

  * Limit production builds only for jessie, trusty and xenial

 _Wed, 18 May 2016 10:40:03 +0300_

---

leicht (0.2-6)

  * Add ca-certificates to build dependencies

 _Wed, 13 Apr 2016 16:12:18 +0300_

---

leicht (0.2-5)

  * Testing build in Travis CI

 _Tue, 12 Apr 2016 21:49:59 +0300_

---

leicht (0.2-4)

  * Use ruby-ronn to generate manpage

 _Mon, 11 Apr 2016 11:07:48 +0300_

---

leicht (0.2-3)

  * Upgrade telegram library

 _Wed, 20 Jan 2016 12:46:50 +0200_

---

leicht (0.2-2)

  * leicht.target now depends on multi-user.target

 _Wed, 20 Jan 2016 11:41:40 +0200_

---

leicht (0.2-1)

  * Release 0.2
  * Postinstall script added
  * Golang library added
  * Manpage updated
  * Improved logging
  * Removing socket node at start(if exists)

 _Wed, 20 Jan 2016 10:52:18 +0200_

---

leicht (0.1-6)

  * postinstall added

 _Fri, 13 Nov 2015 02:10:42 +0200_

---

leicht (0.1-5)

  * Don't panic on non-essential errors
  * Fix error handling
  * Exclude unused fmt package

 _Thu, 12 Nov 2015 12:33:24 +0200_

---

leicht (0.1-4)

  * Fix file existence check

 _Thu, 12 Nov 2015 12:21:06 +0200_

---

leicht (0.1-3)

  * Remove socket before start

 _Thu, 12 Nov 2015 11:42:03 +0200_

---

leicht (0.1-2)

  * Manpage typo fix

 _Wed, 11 Nov 2015 19:25:52 +0200_

---

leicht (0.1-1)

  * Initial release for Debian

 _Sat, 07 Nov 2015 00:15:16 +0200_


src/lib/bash/leicht.sh
function leicht_send_message() {
    chat_id=$1
    reply_id=$2
    message_text=`echo "$3" | jq -R -s '.'`
    disable_web_preview=$4
    leicht_socket=$5

    echo '{ "actionType": "SendMessage", "actionSettings": {"chatID": '$chat_id', "replyToMessageID": '$reply_id', "text": '$message_text', "disableWebPagePreview": '$disable_web_preview'}}' | socat stdio unix-connect:$5
}

function leicht_send_message_to_channel() {
    channel_username=$1
    reply_id=$2
    message_text=`echo "$3" | jq -R -s '.'`
    disable_web_preview=$4
    leicht_socket=$5

    echo '{ "actionType": "SendMessage", "actionSettings": {"channelUsername": "'$channel_username'", "replyToMessageID": '$reply_id', "text": '$message_text', "disableWebPagePreview": '$disable_web_preview'}}' | socat stdio unix-connect:$5
}



src/lib/golang/leicht.go
package leicht

import (
    "net"
    "encoding/json"
    "github.com/Syfaro/telegram-bot-api"
)

type action struct {
    ActionType string           `json:"actionType"`
    ActionSettings interface{}  `json:"actionSettings"`
}

func sendBytesToSocket(bytes []byte, socket string) (err error) {
    conn, err := net.DialUnix("unix", nil, &net.UnixAddr{socket, "unix"})
    if err != nil {
        return err
    }

    _, err = conn.Write(bytes)
    if err != nil {
        return err
    }

    conn.Close()

    return nil
}

func SendMessage(chatID, replyToMessageID int, text string, disableWebPagePreview bool, socket string) (err error) {
    msg := tgbotapi.MessageConfig{
        ChatID: chatID,
        ReplyToMessageID: replyToMessageID,
        Text: text,
        DisableWebPagePreview: disableWebPagePreview,
    }

    act := action{
        ActionType: "SendMessage",
        ActionSettings: msg,
    }

    b, err := json.Marshal(act)
    if err != nil {
        return err
    }

    err = sendBytesToSocket(b, socket)
    if err != nil {
        return err
    }

    return nil
}


func SendMessageToChannel(channelUsername string, replyToMessageID int, text string, disableWebPagePreview bool, socket string) (err error) {
    msg := tgbotapi.MessageConfig{
        ChannelUsername: channelUsername,
        ReplyToMessageID: replyToMessageID,
        Text: text,
        DisableWebPagePreview: disableWebPagePreview,
    }

    act := action{
        ActionType: "SendMessage",
        ActionSettings: msg,
    }

    b, err := json.Marshal(act)
    if err != nil {
        return err
    }

    err = sendBytesToSocket(b, socket)
    if err != nil {
        return err
    }

    return nil
}


src/lib/node.js/leicht.js
exports.sendMessage = function(chatID, replyToMessageID, text, disableWebPagePreview, socket) {
	var net = require('net');
	var action = {
		actionType: "SendMessage",
		actionSettings: {
			chatID: chatID,
			replyToMessageID: replyToMessageID,
			text: text,
			disableWebPagePreview: disableWebPagePreview,
			socket: socket
		}
	}
	
	var client = net.connect({path: socket},
		function() {
			client.write(JSON.stringify(action));
		});
}
exports.sendMessageToChannel = function(channelUsername, replyToMessageID, text, disableWebPagePreview, socket) {
	var net = require('net');
	var action = {
		actionType: "SendMessage",
		actionSettings: {
			channelUsername: channelUsername,
			replyToMessageID: replyToMessageID,
			text: text,
			disableWebPagePreview: disableWebPagePreview,
			socket: socket
		}
	}
	
	var client = net.connect({path: socket},
		function() {
			client.write(JSON.stringify(action));
		});
}


src/leicht/main.go
package main

import (
  "encoding/json"
  goopt "github.com/droundy/goopt"
  "gopkg.in/go-telegram-bot-api/telegram-bot-api.v4"
  "log"
  "net"
  "os"
  "os/exec"
  "os/signal"
  "strconv"
  "syscall"
  "time"
)

var param_cfgpath = goopt.String([]string{"-c", "--config"}, "/etc/leicht/default.json", "set config file path")

func usernameInWhitelist(username string, whitelist []string) bool {
  present := false
  for _, item := range whitelist {
    if username == item {
      present = true
    }
  }
  return present
}

func writeMessageToLog(chatID int64, logDirectory, botUsername, messageText string)  {
  logFilename := ""
  if chatID > 0 {
    logFilename = logDirectory +
      "/user-" +
      strconv.FormatInt(chatID, 10) +
      ".log"
  } else {
    logFilename = logDirectory +
      "/group" +
      strconv.FormatInt(chatID, 10) +
      ".log"
  }

  file, err := os.OpenFile(logFilename, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644)
  if err != nil {
    log.Printf("Could not open log file: %s\n", err.Error())
  }
  _, err = file.WriteString("[" + time.Now().Format("2006-01-02T15:04:05-07:00") + "] <" + botUsername + "> " + messageText + "\n")
  if err != nil {
    log.Printf("Could not write log file entry: %s\n", err.Error())
  }
  file.Close()
}

func main() {

  goopt.Description = func() string {
    return "Leicht - universal telegram bot"
  }

  goopt.Version = "0.1"
  goopt.Summary = "leicht -c [config]"
  goopt.Parse(nil)

  CfgParams, _ := LoadConfig(*param_cfgpath)

  msgbus := make(chan string)

  bot, err := tgbotapi.NewBotAPI(CfgParams.Token)
  if err != nil {
    log.Panic(err)
  }

  bot.Debug = CfgParams.Debug

  log.Printf("Authorized on account %s", bot.Self.UserName)

  u := tgbotapi.NewUpdate(0)
  u.Timeout = 60

  updates, err := bot.GetUpdatesChan(u)
  if err != nil {
    log.Panic(err)
  }


  go func() {
    for update := range updates {
      var MessageID int
      var MessageText string
      var UserID string
      var ChatID int64

      if update.Message != nil {
        MessageID = update.Message.MessageID
        MessageText = update.Message.Text
        UserID = update.Message.From.UserName
        ChatID = update.Message.Chat.ID
      } else if update.ChannelPost != nil {
        MessageID = update.ChannelPost.MessageID
        MessageText = update.ChannelPost.Text
        UserID = update.ChannelPost.Chat.Title
        ChatID = update.ChannelPost.Chat.ID
      }

      if CfgParams.Logging {
        if CfgParams.WhitelistEnabled {
          if !CfgParams.DoNotLogBlacklisted || usernameInWhitelist(UserID, CfgParams.Whitelist) {
            writeMessageToLog(ChatID,
                              CfgParams.LogDirectory,
                              UserID,
                              MessageText)
          }
        } else {
          writeMessageToLog(ChatID,
                            CfgParams.LogDirectory,
                            UserID,
                            MessageText)
        }
      }

      if !CfgParams.WhitelistEnabled || usernameInWhitelist(UserID, CfgParams.Whitelist) {
        var args []string
        args = append(args,
          UserID,
          strconv.FormatInt(ChatID, 10),
          strconv.Itoa(MessageID))
        if update.Message.Voice != nil {
          url, err := bot.GetFileDirectURL(update.Message.Voice.FileID)
          if err != nil {
            log.Printf("Could not get direct URL of a voice file: %s\n", err.Error())
          }
          args = append(args, "handle_voice "+url)
        } else {
          args = append(args, MessageText)
        }
        cmd := exec.Command(CfgParams.Script, args...)
        err := cmd.Start()
        if err != nil {
          log.Printf("Could not execute command: %s\n", err.Error())
        }
        logstring := make([]interface{}, len(args)+1)
        logstring[0] = CfgParams.Script
        for i := range args {
          logstring[i+1] = args[i]
        }
        log.Printf("Executing script: %s %s %s %s \"%s\"\n", logstring...)
      }
    }
  }()

  go func() {
    if _, err := os.Stat(CfgParams.Socket); err == nil {
      log.Printf("Socket %s exists! Removing...\n", CfgParams.Socket)
      os.Remove(CfgParams.Socket)
    }
    l, err := net.ListenUnix("unix", &net.UnixAddr{CfgParams.Socket, "unix"})
    if err != nil {
      log.Fatalf("Could not open UNIX socket: %s\n", err.Error())
    }
    err = os.Chmod(CfgParams.Socket, CfgParams.SocketMode)
    if err != nil {
      log.Fatalf("Could not change permissions for UNIX socket: %s\n", err.Error())
    }
    for {
      conn, err := l.AcceptUnix()
      if err != nil {
        log.Fatalf("Could not accept data from UNIX socket: %s\n", err.Error())
      }
      var buf [1024]byte
      n, err := conn.Read(buf[:])
      if err != nil {
        log.Fatalf("Could not read from UNIX socket: %s\n", err.Error())
      }
      msgbus <- string(buf[:n])
    }
  }()

  c := make(chan os.Signal, 1)
  signal.Notify(c, syscall.SIGINT,
    syscall.SIGTERM,
    syscall.SIGQUIT)
  go func() {
    for sig := range c {
      os.Remove(CfgParams.Socket)
      log.Printf("Captured %v, Exiting\n", sig)
      os.Exit(0)
    }
  }()

  for {
    msg := <-msgbus
    var action map[string]*json.RawMessage
    data := []byte(msg)
    err := json.Unmarshal(data, &action)
    if err != nil {
      log.Printf("Could not unmarshal incoming JSON: %s\n", err.Error())
    } else {
      var actionType string
      err = json.Unmarshal(*action["actionType"], &actionType)
      if err != nil {
        log.Printf("Could not detect action type: %s\n", err.Error())
      } else {
        switch actionType {
        case "SendMessage":
          var settings tgbotapi.MessageConfig
          err = json.Unmarshal(*action["actionSettings"], &settings)
          if err != nil {
            log.Printf("Could not get action settings: %s\n", err.Error())
          } else {
            bot.Send(settings)
            if CfgParams.Logging {
              writeMessageToLog(settings.ChatID,
                                CfgParams.LogDirectory,
                                bot.Self.UserName,
                                settings.Text)
            }
          }
        }
      }
    }
  }
}
